File: tests/test_scoring.py
import unittest

from painradar.models import Post
from painradar.clustering import Cluster
from painradar.scoring import score_cluster


def make_post(id_, author, score=10, num_comments=5, subreddit="sub1", created_utc=1):
    return Post(
        id=id_,
        title=f"title {id_}",
        selftext="I manually do this every week, so frustrating, expensive tools, I'd pay for a fix.",
        subreddit=subreddit,
        author=author,
        score=score,
        num_comments=num_comments,
        created_utc=created_utc,
        url=f"https://reddit.com/{id_}",
    )


class TestScoreCluster(unittest.TestCase):
    def test_more_independent_authors_scores_higher(self):
        low = Cluster(posts=[make_post("1", "a"), make_post("2", "a")])
        high = Cluster(posts=[make_post("3", "a"), make_post("4", "b")])
        low_result = score_cluster(low)
        high_result = score_cluster(high)
        self.assertGreater(high_result.total_score, low_result.total_score)

    def test_more_subreddits_scores_higher(self):
        one_sub = Cluster(posts=[make_post("1", "a", subreddit="s1"), make_post("2", "b", subreddit="s1")])
        two_subs = Cluster(posts=[make_post("3", "a", subreddit="s1"), make_post("4", "b", subreddit="s2")])
        self.assertGreater(score_cluster(two_subs).total_score, score_cluster(one_sub).total_score)

    def test_higher_engagement_scores_higher(self):
        low_engagement = Cluster(posts=[make_post("1", "a", score=1, num_comments=0)])
        high_engagement = Cluster(posts=[make_post("2", "a", score=500, num_comments=200)])
        self.assertGreater(
            score_cluster(high_engagement).total_score, score_cluster(low_engagement).total_score
        )

    def test_score_result_has_explainable_reasons(self):
        cluster = Cluster(posts=[make_post("1", "a"), make_post("2", "b")])
        result = score_cluster(cluster)
        self.assertIsInstance(result.reasons, list)
        self.assertTrue(len(result.reasons) > 0)
        self.assertTrue(all(isinstance(r, str) for r in result.reasons))

    def test_score_is_bounded(self):
        cluster = Cluster(posts=[make_post(str(i), f"author{i}", score=99999, num_comments=99999) for i in range(20)])
        result = score_cluster(cluster)
        self.assertLessEqual(result.total_score, 100.0)
        self.assertGreaterEqual(result.total_score, 0.0)

    def test_market_evidence_state_is_reported(self):
        cluster = Cluster(posts=[make_post("1", "a"), make_post("2", "b")])
        result = score_cluster(cluster)
        self.assertIn(result.market_evidence_state, {"faible", "modérée", "forte"})


if __name__ == "__main__":
    unittest.main()

File: test/test_scoring.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import { Post } from "../src/models.js";
import { Cluster } from "../src/clustering.js";
import { scoreCluster } from "../src/scoring.js";

function makePost(
  id: string,
  author: string,
  opts: Partial<Post> = {},
): Post {
  return {
    id,
    title: `title ${id}`,
    selftext: "I manually do this every week, so frustrating, expensive tools, I'd pay for a fix.",
    subreddit: "sub1",
    author,
    score: 10,
    num_comments: 5,
    created_utc: 1,
    url: `https://reddit.com/${id}`,
    ...opts,
  };
}

test("more independent authors scores higher", () => {
  const low = new Cluster([makePost("1", "a"), makePost("2", "a")]);
  const high = new Cluster([makePost("3", "a"), makePost("4", "b")]);
  assert.ok(scoreCluster(high).totalScore > scoreCluster(low).totalScore);
});

test("more subreddits scores higher", () => {
  const oneSub = new Cluster([
    makePost("1", "a", { subreddit: "s1" }),
    makePost("2", "b", { subreddit: "s1" }),
  ]);
  const twoSubs = new Cluster([
    makePost("3", "a", { subreddit: "s1" }),
    makePost("4", "b", { subreddit: "s2" }),
  ]);
  assert.ok(scoreCluster(twoSubs).totalScore > scoreCluster(oneSub).totalScore);
});

test("higher engagement scores higher", () => {
  const low = new Cluster([makePost("1", "a", { score: 1, num_comments: 0 })]);
  const high = new Cluster([makePost("2", "a", { score: 500, num_comments: 200 })]);
  assert.ok(scoreCluster(high).totalScore > scoreCluster(low).totalScore);
});

test("negative engagement values cannot make the score non-finite", () => {
  const cluster = new Cluster([
    makePost("1", "a", { score: -100, num_comments: 5 }),
    makePost("2", "b", { score: 2, num_comments: -50 }),
  ]);

  const result = scoreCluster(cluster);

  assert.ok(Number.isFinite(result.totalScore));
  assert.ok(result.totalScore >= 0 && result.totalScore <= 100);
  assert.ok(result.reasons.includes("Engagement cumulé (score + commentaires) : 7."));
});

test("score result has explainable reasons", () => {
  const cluster = new Cluster([makePost("1", "a"), makePost("2", "b")]);
  const result = scoreCluster(cluster);
  assert.ok(Array.isArray(result.reasons));
  assert.ok(result.reasons.length > 0);
  assert.ok(result.reasons.every((r) => typeof r === "string"));
});

test("score is bounded 0..100", () => {
  const posts = Array.from({ length: 20 }, (_, i) =>
    makePost(String(i), `author${i}`, { score: 99999, num_comments: 99999 }),
  );
  const result = scoreCluster(new Cluster(posts));
  assert.ok(result.totalScore <= 100.0);
  assert.ok(result.totalScore >= 0.0);
});

test("market evidence state is reported", () => {
  const cluster = new Cluster([makePost("1", "a"), makePost("2", "b")]);
  const result = scoreCluster(cluster);
  assert.ok(["faible", "modérée", "forte"].includes(result.marketEvidenceState));
});
